model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java

```diff
@@ -1016,40 +1017,70 @@ public Stream<GroupModel> getGroupsByRoleStream(RealmModel realm, RoleModel role
 
     @Override
     public Stream<GroupModel> getTopLevelGroupsStream(RealmModel realm, String search, Boolean exact, Integer first, Integer max) {
-        String cacheKey = getTopGroupsQueryCacheKey(realm.getId() + search + first + max);
-        boolean queryDB = invalidations.contains(cacheKey) || listInvalidations.contains(cacheKey)
+        // TODO the way the cache is implemented, we can't invalidate keys based on only one property matching
+        // to take advantage of the cache we have to load all groups and then paginate/filter the stream
+        String cacheKey = getTopGroupsQueryCacheKey(realm.getId());
+
+        boolean isInvalid = invalidations.contains(cacheKey) || listInvalidations.contains(cacheKey)
             || listInvalidations.contains(realm.getId());
-        if (queryDB) {
-            return getGroupDelegate().getTopLevelGroupsStream(realm, search, exact, first, max);
-        }
+
+//        if (queryDB) {
+//            return getGroupDelegate().getTopLevelGroupsStream(realm);
+//        }
 
         GroupListQuery query = cache.get(cacheKey, GroupListQuery.class);
+
         if (Objects.nonNull(query)) {
             logger.tracev("getTopLevelGroups cache hit: {0}", realm.getName());
         }
 
+        List<GroupModel> model;
```

Comment:
Here we can maybe safe one stream consumption when the is invalidated or cache hit:

```Java
Stream<GroupModel> groupsStream;
 // cache miss so query db and populate the cache
if (Objects.isNull(query)) {
    Long loaded = cache.getCurrentRevision(cacheKey);
    List<GroupModel> model = getGroupDelegate().getTopLevelGroupsStream(realm).collect(Collectors.toList());
    if (model.isEmpty()) return Stream.empty();
    if (!isInvalid) {
        Set<String> ids = new HashSet<>();
        for (GroupModel group : model) ids.add(group.getId());
        query = new GroupListQuery(loaded, cacheKey, realm, ids);
        logger.tracev("adding realm getTopLevelGroups cache miss: realm {0} key {1}", realm.getName(), cacheKey);
        cache.addRevisioned(query, startupRevision);
    }
    groupsStream = model.stream();
} else if (isInvalid) {
    // cache invalid so get our groups from the db
    groupsStream = getGroupDelegate().getTopLevelGroupsStream(realm);
} else {
    // cache hit so let's load from the cache
    Set<String> groupIds = query.getGroups();
    groupsStream = groupIds.stream().map(groupId -> getGroupById(realm, groupId)).filter(Objects::nonNull);
}

Stream<GroupModel> groups = groupsStream.sorted(GroupModel.COMPARE_BY_NAME);
```